src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,405 @@
+      return ReplaceExprByVar(src1);
+    }
+
+    if (const auto* node1 = src1.as<TupleNode>()) {
+      if (const auto* node2 = src2.as<TupleNode>()) {
+        ICHECK(node1->fields.size() == node2->fields.size()) << "size of tuple not match";
+        Array<Expr> result;
+        for (size_t i = 0; i < node1->fields.size(); ++i) {
+          result.push_back(DoAdd(node1->fields[i], node2->fields[i]));
+        }
+        return Tuple(result);
+      } else {
+        LOG(FATAL) << "Type not match: src1 and src2 should be both tuple";
+        return Expr();
+      }
+    } else {
+      return Call(add_op_, {src1, ReplaceExprByVar(src2)});
+    }
+  }
+
+  Tuple DoAddInTuple(const Tuple& origin, int index, const Expr& increment) {
+    Array<Expr> ret;
+    for (size_t i = 0; i < origin->fields.size(); ++i) {
+      if ((int)i == index) {
+        ret.push_back(DoAdd(origin->fields[i], increment));
+      } else {
+        ret.push_back(origin->fields[i]);
+      }
+    }
+    return Tuple(ret);
+  }
+
+  void BindAndEmit(Var v, Expr e) {
+    if (adjoint_expr_to_var_.count(e)) {
+      e = adjoint_expr_to_var_[e];
+    } else {
+      adjoint_expr_to_var_.Set(e, v);
+      // e->checked_type_ = v->checked_type();
+      // e->shape_ = v->shape();
+    }
+    if (v->IsInstance<DataflowVarNode>()) {
+      builder_->Emit(VarBinding(v, e));
+    } else {
+      builder_->EmitOutput(VarBinding(v, e));
+    }
+  }
+
+  void CheckTarget(const Expr& e) {
+    ICHECK(!e->IsInstance<DataflowVarNode>()) << "not an output node";
+    ICHECK(e->checked_type().as<DynTensorTypeNode>()) << "target must be a DynTensorType";
+    ICHECK(e->shape().as<ShapeExprNode>()) << "error when getting target shape";
+    const auto* shape_node = e->shape().as<ShapeExprNode>();
+    ICHECK(shape_node->values.size() == 0) << "target must be a scalar";
+  }
+
+  void InitGrad(const Var& adjoint_var, const Var& var) {
+    ObjectPtr<InitAttrs> attrs = make_object<InitAttrs>();
+    auto type = Downcast<DynTensorType>(var->checked_type());
+    attrs->dtype = type->dtype;
+
+    const Expr& init = Call(ones_op_, {var->shape()}, Attrs(attrs));
+    adjoint_expr_map_.Set(var, init);
+  }
+
+  // the differentiation target
+  Var target_var_;
+  // the arguments to differentiate
+  Array<Var> require_grads_;
+
+  // var to its adjoints var
+  Map<Var, Var> adjoint_var_map_;
+  // var to its adjoint expr
+  Map<Var, Expr> adjoint_expr_map_;
+  // trace binding
+  Map<Expr, Var> adjoint_expr_to_var_;
+  // track zeros introduced
+  std::unordered_set<Expr, ObjectPtrHash, ObjectPtrEqual> zeros_tracker_;
+
+  // gop map
+  const OpAttrMap<FPrimalGradient> gradient_op_map_ =
+      Op::GetAttrMap<FPrimalGradient>("FPrimalGradient");
+
+  // constant
+  const Op& ones_op_ = Op::Get("relax.ones");
+  const Op& add_op_ = Op::Get("relax.add");
+  const Op& zeros_op_ = Op::Get("relax.zeros");
+};
+
+/*!
+ * \brief A simple reverse-mode auto differentiation.
```

Comment:
I’m thinking about providing more documentation. Feel free to rephrase my words if you have better ones.
```suggestion
 * \brief A simple reverse-mode auto differentiation. For a given function specified by the input global var, it generates a new function which computes the adjoints of the given function output on the given inputs.
```

Reply:
What about handling documents like this:
- In `src/relax/transform/gradient.cc`, we only provide succinct file document;
- in `include/tvm/relax/transform.h`, we provide detailed document for pass;
- in `python/tvm/relax/transform/transform.py`, we provide detailed document for python api.

Writing document in headers is more friendly for doxygen.

Reply:
Sounds good. In headers or on Python side you want to explain the API clearly, and on the impl side you can tell more about the impl algorithms if you want. In this specific place I think it’s good to remove the doc. Just make sure that whenever we provide documents, we want to provide clear ones.